### ISSUE TITLE
Discussion issue for sandbox permissions / --fix-perms

### ISSUE THREAD

**Issue by dtrudg** [apptainer/singularity#4671]
_This issue is referred to in Singularity's `WARNING` output and should be kept open for discussion_

----

## TLDR

Singularity 3.5 does not adjust permissions of files extracted from docker/oci sources. This is a change from 3.4 and `singularity build --sandbox mysandbox docker://centos:7` will now create a sandbox which cannot be removed without `chmod -R u+rwX` or similar.

A `--fix-perms` flag was added to preserve the behaviour of 3.4, mangling permissions of extracted files.

~~We have proposed to deprecate the `--fix-perms` option in 3.6, and remove it in 3.8 - subject to discussion here.~~

## Background

Singularity has supported building from Docker / OCI container images for some time now, but it has never been truly faithful to the content of those images. Until now, when you build a container from a Docker / OCI source, Singularity has explicitly modified permissions, so that the creator always has at least read/write on files and read/write/execute on directories. This has applied whether the destination is a SIF file, or a sandbox directory. When a SIF image is created, ownership becomes 'root' in a read-only squashfs, but for a sandbox the user creator can always modify any file/directory in the sandbox directly, and possibly accidentally.

Modifying the permissions in this way is a little ugly. It means that Singularity containers differ from the intent of the original Docker / OCI image. As we move forward in a world which now includes CRI support to use Singularity under Kubernetes, and the increasing use of other rootless tools that deal with OCI containers, faithful adherence to the permissions in Docker / OCI images is an important goal.

## Change in 3.5

For Singularity 3.5 some changes have been made that allow us to be faithful to source image permissions. If you are building to SIF there is no real impact to users. If you are building to a sandbox from Ubuntu or Debian base images there is no impact. However, there is an impact for users building to a sandbox, from CentOS base images, or others with layers containing files that have restrictive permissions.

If you build such an image into a sandbox , Singularity will warn you that you will need to `chmod` to remove:

```
WARNING: Permission handling has changed in Singularity 3.5 for improved OCI compatibility
WARNING: The sandbox will contain files/dirs that cannot be removed until permissions are modified
WARNING: Use 'chmod -R u+rwX' to set permissions that allow removal
WARNING: Use the '--fixperms' option to 'singularity build' to modify permissions at build time
WARNING: You can provide feedback about this change at https://github.com/sylabs/singularity/issues/4671
```

Note that a new option `--fix-perms` is mentioned. This restores the behaviour of 3.4, with a warning that permissions are being modified ~~and informs about the deprecation plan / this discussion~~:

```
WARNING: The --fix-perms option modifies the filesystem permissions on the resulting container.
```

## Proposals / Discussion

~We propose that the `--fix-perms` option will be marked as deprecated in Singularity 3.6, 3.7, and then removed from Singularity 3.8. The `--fix-perms` option is creating a container that is not faithful to the original image, and a `chmod` command is sufficient to handle sandboxes created faithfully.~

We seek feedback on:

1. How often you are impacted (i.e. perform builds of CentOS/RHEL derived containers to sandbox)
2. How much impact the need for `chmod` or `--fix-perms` causes to you / your users.
~3. Whether the proposal to remove `--fix-perms` in future seems reasonable to you.~

**Comment by gmkurtzer** [comment]
Great write up, thanks @dctrud!

1. I've seen the problem come up numerous times with multiple containers (which was why I originally added that code ;-))
2. In my experience, it will cause some confusion, but a good warning will help.
3. I don't believe that the `--fix-perms` options should have a planned deprecation. 

Greg

**Comment by dtrudg** [comment]
Thanks for commenting @gmkurtzer. We are looking at an RC on Wednesday for 3.5, and then a final release within 2 weeks of that so there is time to drop the deprecation plan and warning of `--fix-perms` or make other adjustments if needed.

> I've seen the problem come up numerous times with multiple containers (which was why I originally added that code ;-))

I think I've now coded the same thing for 3 different extraction methods that superseded the original code... `libarchive` based in 2.x, `containers/image` in 3.0, `umoci` in 3.4. Somehow OCI permissions handling has become my perpetual task :-)

**Comment by DrDaveD** [comment]
I agree with @gmkurtzer that --fix-perms should be a permanent option.

**Comment by dtrudg** [comment]
Okay - looks like we'll drop the deprecation plans. I'll make the changes for that in the messages for an `rc.2` next week. Cheers.

**Comment by dtrudg** [comment]
PR #4694 which deletes the deprecation/removal warning for `--fix-perms` is proposed for `3.5.1-rc.2`  and I've struck out the references to deprecation/removal here.

This issue remains for comment on the effect of the permission handling change.

**Comment by dtrudg** [comment]
There has been no comment here. I am going to close the issue.
